formula/src/main/java/com/instacart/formula/internal/Listeners.kt

```diff
@@ -2,31 +2,29 @@ package com.instacart.formula.internal
 
 internal class Listeners {
     private var listeners: SingleRequestMap<Any, ListenerImpl<*, *, *>>? = null
+    private var indexes: MutableMap<Any, Int>? = null
 
     private fun duplicateKeyErrorMessage(key: Any): String {
-        if (key is String) {
-            // This indicates manual key creation.
-            return "Listener $key is already defined. Make sure your key is unique."
-        }
-        // This indicates automatic key generation
-        return "Listener $key is already defined. Are you calling it in a loop or reusing a method? You can wrap the call with FormulaContext.key"
+        return "Listener $key is already defined. Unexpected issue."
     }
 
     fun <Input, State, Event> initOrFindCallback(key: Any): ListenerImpl<Input, State, Event> {
-        val listeners = listeners ?: run {
-            val initialized: SingleRequestMap<Any, ListenerImpl<*, *, *>> = mutableMapOf()
-            this.listeners = initialized
-            initialized
+        val currentHolder = listenerHolder<Input, State, Event>(key)
+        return if (currentHolder.requested) {
+            val index = nextIndex(key)
+            val indexedKey = IndexedKey(key, index)
+            initOrFindCallback(indexedKey)
```

Comment:
Maybe add a check here that `key != indexedKey`, to avoid possible endless recursion?

Reply:
I might be missing something, but I don't think it's possible for them to ever be equal.

Reply:
Added an extra check